### DISCUSSION TITLE
Best practices in regex matching and editing the entire component content

### DISCUSSION THREAD

**Discussion by kangarko** [KyoriPowered/adventure#1094] [General]
I have an anti-caps feature in my plugin and I am looking to know what is the best practice to replace excessive caps in an Adventure component.

Current code (chatMessage is a legacy string):

```java
final String[] words = chatMessage.split(" ");

boolean capsAllowed = false;
boolean whitelisted = false;

for (int i = 0; i < words.length; i++) {
	final String word = words[i];

       // Some logic code here
	 
	if (!whitelisted) {
		if (!capsAllowed) {
			final char firstChar = word.charAt(0);

			words[i] = firstChar + word.toLowerCase().substring(1);

		} else
			words[i] = word.toLowerCase();

		capsAllowed = !words[i].endsWith(".") && !words[i].endsWith("!") && !words[i].endsWith("?");
	}

	whitelisted = false;
}

// I need to set the component here.
String.join(" ", words);
```

My current solution involves converting the component to legacy and then back which destroys hover/click events and to my knowledge has a unnecessarily performance hit. 

Any advice would be greatly appreciated!

Matej

**Comment by PedroMPagani** [comment]
A couple of rewrites and you can get there
